File: Day-07/python/Schloool/day_07.py
input = open('input.txt').read().splitlines()

class Dir:
    def __init__(self, name, parent):
        self.name = name
        self.parent = parent
        self.subdirs = {}
        self.files = []

root = Dir('/', None)
current_dir = root

for line in input[1:]:
    line_parts = line.split(' ')
    if line.startswith('dir'):
        current_dir.subdirs[line_parts[1]] = Dir(line_parts[1], current_dir)
    if line.startswith('$ cd'):
        current_dir = current_dir.parent if line_parts[2] == '..' else current_dir.subdirs[line_parts[2]]
    if line_parts[0].isnumeric():
        current_dir.files.append(int(line_parts[0]))

dir_sizes = []
def calc_dir_sizes(dir):
    file_size = sum(dir.files)
    for subdir in dir.subdirs.values():
        file_size += calc_dir_sizes(subdir)
    dir_sizes.append(file_size)
    return file_size

# Task 1
calc_dir_sizes(root)
print(sum([size for size in dir_sizes if size < 100_000]))

# Task 2
needed_space = 30_000_000 - (70_000_000 - dir_sizes[-1])
print(min([size for size in dir_sizes if size >= needed_space]))

File: Day-07/nim/Duckulus/day07.nim
import strutils, os

type File = object
  name: string
  size: int

type FileTreeNode = ref object
  name: string
  files: seq[File]
  children: seq[FileTreeNode]
  parent: FileTreeNode

proc newNode(name: string): FileTreeNode=
  var node = new(FileTreeNode)
  node.name= name
  node.files= @[]
  node.children= @[]
  return node

proc addChild(node: var FileTreeNode, name: string): FileTreeNode =
  var n = newNode(name)
  n.parent = node
  node.children.add(n)
  return n

proc addFile(node: var FileTreeNode, file: File)=
  node.files.add(file)

proc totalSize(node: FileTreeNode):int=
  var sum = 0
  for f in node.files:
    sum+=f.size
  for n in node.children:
    sum+=totalSize(n)
  return sum

proc printFileTree(root: FileTreeNode, depth: int = 0): void=
  echo " ".repeat(depth),"-", root.name
  for f in root.files:
    echo " ".repeat(depth+1), f
  for n in root.children:
    printFileTree(n, depth+4)

var root = newNode("root")
var currentNode: FileTreeNode

if os.paramCount()<1:
  echo "No input file specified"
  quit(1)

var lines = readFile(paramStr(1)).strip.splitLines

var i = 0
while i<lines.len:
  let line = lines[i]
  if(line.startsWith("$")):
    let command = line.split(" ")[1]
    case command:
      of "cd":
        let arg = line.split(" ")[2]
        case arg:
          of "/":
            currentNode = root
          of "..":
            currentNode = currentNode.parent
          else:
            currentNode = currentNode.addChild(arg)
  else:
    if(not line.startsWith("dir")):
      currentNode.addFile(File(name: line.split(" ")[1], size: line.split(" ")[0].parseInt))
  i+=1

#printFileTree(root)

block part1:
  var sum = 0
  var nodes = @[root]
  while(nodes.len>0):
    var node = nodes[0]
    nodes.delete(0)
    var size = node.totalSize
    if(size<100000):
      sum+=size
    for child in node.children:
      nodes.add(child)
  echo "Part 1: ", sum

block part2:
  var unusedSpace = 70000000 - root.totalSize
  var missingSpace = 30000000 - unusedSpace
  var smallest = root.totalSize
  var nodes = @[root]
  while(nodes.len>0):
    var node = nodes[0]
    nodes.delete(0)
    var size = node.totalSize
    if(size>missingSpace and size<smallest):
      smallest = size
    for child in node.children:
      nodes.add(child)
  echo "Part 2: ", smallest